src/OpenTelemetry.Api/Context/CorrelationContext.cs

```diff
@@ -25,46 +26,73 @@ namespace OpenTelemetry.Context
     /// </summary>
     public readonly struct CorrelationContext : IEquatable<CorrelationContext>
     {
-        private static readonly List<CorrelationContextEntry> EmptyList = new List<CorrelationContextEntry>();
-        private readonly List<CorrelationContextEntry> entries;
+        internal static readonly CorrelationContext Empty = new CorrelationContext(null);
+        internal static readonly IEnumerable<KeyValuePair<string, string>> EmptyBaggage = new KeyValuePair<string, string>[0];
+        private readonly Activity activity;
+
+        internal CorrelationContext(in Activity activity)
+        {
+            this.activity = activity;
+        }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="CorrelationContext"/> struct.
+        /// Gets the current <see cref="CorrelationContext"/>.
         /// </summary>
-        /// <param name="entries">Entries for correlation context.</param>
-        internal CorrelationContext(List<CorrelationContextEntry> entries)
+        public static CorrelationContext Current
         {
-            this.entries = entries;
+            get
+            {
+                Activity activity = Activity.Current;
+                return activity == null
+                    ? Empty
+                    : new CorrelationContext(activity);
+            }
         }
 
         /// <summary>
-        /// Gets empty object of <see cref="CorrelationContext"/> struct.
+        /// Gets the correlation values.
         /// </summary>
-        public static CorrelationContext Empty { get; } = new CorrelationContext(EmptyList);
+        public IEnumerable<KeyValuePair<string, string>> Correlations => this.activity?.Baggage ?? EmptyBaggage;
```

Comment:
Couple questions:
* Would there be memory allocation concern?
* What would happen if a consumer gets the IEnumerable and tries to iterate through it, while the underlying baggage got modified?
* Does this enforce immutability? https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/correlationcontext/api.md#get-correlations

Reply:
> Would there be memory allocation concern?

Check out the underlying Activity.Baggage enumeration:
https://github.com/dotnet/runtime/blob/ead2cd50799ecc523e23f50e50b54d8a30d0aa4f/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L290

Right below that method is the get.

The memory won't be too bad. An allocation for the state thing generated by the compiler due to the `yield`. The internal storage is a LinkedList, not great for searching. O(N)?

> What would happen if a consumer gets the IEnumerable and tries to iterate through it, while the underlying baggage got modified?

I think the way the LinkedList works under the hood, it would actually be OK. User can only add to the back of the list. So if you were enumerating you would also consume that new one when you got to the tail.

> Does this enforce immutability?

I think so. The enumeration is returning the KVPs. They are structs, so you essentially get a copy?